File: frontend/lib/api/vault.ts
import { invoke } from "./client";

/**
 * Vault entry create/update/delete/list/status IPC wrappers.
 *
 * The shape of `VaultEntrySafe` (and its `type` enum) mirrors
 * `golish-core/src/vault.rs::VaultEntrySafe` — the canonical Rust
 * struct returned by all `vault_*` Tauri commands. Field names are
 * snake_case to match the JSON over the wire (Rust serializes with
 * `#[serde(rename = "type")]` for `entry_type`).
 *
 * Argument types (`AddVaultEntryParams` etc.) below stay hand-written
 * because Tauri's auto camelCase→snake_case conversion only applies to
 * *arguments* (front → back), and they don't 1:1 mirror any Rust struct.
 *
 * Whenever the Rust struct changes, update the field set / enum literals
 * here in lockstep. (Previously this came from `lib/generated/` ts-rs
 * codegen which is being removed per M2.5.)
 */

export type VaultEntryType =
  | "password"
  | "token"
  | "ssh_key"
  | "api_key"
  | "cookie"
  | "certificate"
  | "other";

export interface VaultEntrySafe {
  id: string;
  name: string;
  type: VaultEntryType;
  username: string;
  notes: string;
  project: string;
  tags: string[];
  status: string;
  source_url: string;
  /**
   * Unix seconds. Nullable when the entry has never been validated.
   * Frontend arithmetic (e.g. `entry.last_validated_at * 1000`) treats
   * this as `number` (timestamps fit in `Number.MAX_SAFE_INTEGER`).
   */
  last_validated_at: number | null;
  created_at: number;
  updated_at: number;
}

export interface AddVaultEntryParams {
  name: string;
  /** "password" | "token" | "api_key" | "cookie" | "other" — keep in sync with `VaultEntryType` enum. */
  entryType: string;
  value: string;
  username?: string | null;
  notes?: string | null;
  /**
   * Host/project the credential belongs to (e.g. `evil.example.com`).
   * Stored as the `project` column in `vault_entries`.
   */
  project?: string | null;
  /** Free-form tags shown in the credential list. */
  tags?: string[] | null;
  /** Origin URL the credential was captured from. */
  sourceUrl?: string | null;
  projectPath: string | null;
}

export async function addVaultEntry(params: AddVaultEntryParams): Promise<VaultEntrySafe> {
  return invoke<VaultEntrySafe>("vault_add", params as unknown as Record<string, unknown>);
}

export async function listVaultEntries(projectPath: string | null): Promise<VaultEntrySafe[]> {
  return invoke<VaultEntrySafe[]>("vault_list", { projectPath });
}

/**
 * Run validation on a stored credential (e.g. attempt a probe with
 * the stored value). Returns a human-readable status string.
 */
export async function validateVaultEntry(id: string, projectPath: string | null): Promise<string> {
  return invoke<string>("vault_validate", { id, projectPath });
}

/**
 * Fetch the plaintext credential value for an entry (one-shot).
 * Triggers an audit log entry server-side.
 */
export async function getVaultValue(id: string, projectPath: string | null): Promise<string> {
  return invoke<string>("vault_get_value", { id, projectPath });
}

export async function deleteVaultEntry(id: string, projectPath: string | null): Promise<void> {
  await invoke("vault_delete", { id, projectPath });
}

export interface UpdateVaultEntryParams {
  id: string;
  /** New plaintext credential value (only required when rotating the secret). */
  value?: string | null;
  /** Updated username (e.g. when a duplicate-host credential changes its login). */
  username?: string | null;
  /** Free-form notes shown in the credential vault UI. */
  notes?: string | null;
  projectPath: string | null;
}

export async function updateVaultEntry(params: UpdateVaultEntryParams): Promise<void> {
  await invoke("vault_update", params as unknown as Record<string, unknown>);
}

export async function updateVaultStatus(
  id: string,
  status: string,
  projectPath: string | null
): Promise<void> {
  await invoke("vault_update_status", { id, status, projectPath });
}

File: backend/crates/golish-core/src/vault.rs
//! Simple XOR-based value obfuscation for credential vault storage.
//!
//! NOT cryptographically secure — provides only basic obfuscation so
//! values are not stored in plaintext on disk.

use base64::{engine::general_purpose::STANDARD as B64, Engine};
use serde::{Deserialize, Serialize};
use ts_rs::TS;

fn derive_key() -> Vec<u8> {
    let seed = format!(
        "golish-vault-{}",
        dirs::home_dir()
            .map(|h| h.to_string_lossy().to_string())
            .unwrap_or_default()
    );
    let mut key = Vec::with_capacity(32);
    let bytes = seed.as_bytes();
    for i in 0..32 {
        key.push(bytes[i % bytes.len()].wrapping_add(i as u8).wrapping_mul(7));
    }
    key
}

pub fn obfuscate(plain: &str) -> String {
    let key = derive_key();
    let encrypted: Vec<u8> = plain
        .as_bytes()
        .iter()
        .enumerate()
        .map(|(i, b)| b ^ key[i % key.len()])
        .collect();
    B64.encode(&encrypted)
}

pub fn deobfuscate(encoded: &str) -> Result<String, String> {
    let key = derive_key();
    let data = B64.decode(encoded).map_err(|e| e.to_string())?;
    let plain: Vec<u8> = data
        .iter()
        .enumerate()
        .map(|(i, b)| b ^ key[i % key.len()])
        .collect();
    String::from_utf8(plain).map_err(|e| e.to_string())
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct VaultEntry {
    pub id: String,
    pub name: String,
    #[serde(rename = "type")]
    pub entry_type: VaultEntryType,
    pub value: String,
    #[serde(default)]
    pub username: String,
    #[serde(default)]
    pub notes: String,
    #[serde(default)]
    pub project: String,
    #[serde(default)]
    pub tags: Vec<String>,
    pub created_at: u64,
    pub updated_at: u64,
}

/// Sanitized vault entry returned to the UI (no plaintext value).
///
/// Field names are deliberately snake_case (matches storage schema) and
/// `entry_type` is exposed as `type` over the wire via `#[serde(rename = "type")]`.
/// The `#[derive(TS)]` + `#[ts(export)]` exports a TypeScript counterpart to
/// `frontend/lib/generated/VaultEntrySafe.ts` so the frontend never has to
/// hand-mirror this shape again. Run `just generate-types` after editing.
#[derive(Debug, Clone, Serialize, Deserialize, TS)]
#[ts(export, export_to = "generated/")]
pub struct VaultEntrySafe {
    pub id: String,
    pub name: String,
    #[serde(rename = "type")]
    #[ts(rename = "type")]
    pub entry_type: VaultEntryType,
    pub username: String,
    pub notes: String,
    pub project: String,
    pub tags: Vec<String>,
    pub status: String,
    pub source_url: String,
    /// Unix seconds. Override the default `bigint` mapping because timestamps
    /// fit comfortably in JS `number` (Number.MAX_SAFE_INTEGER = 2^53 is well
    /// beyond seconds since epoch for the next ~285M years) and the frontend
    /// arithmetic (`entry.created_at * 1000`) treats them as `number`.
    #[ts(type = "number | null")]
    pub last_validated_at: Option<u64>,
    #[ts(type = "number")]
    pub created_at: u64,
    #[ts(type = "number")]
    pub updated_at: u64,
}

#[derive(Debug, Clone, Serialize, Deserialize, PartialEq, TS)]
#[serde(rename_all = "lowercase")]
#[ts(export, export_to = "generated/")]
pub enum VaultEntryType {
    Password,
    Token,
    #[serde(rename = "ssh_key")]
    #[ts(rename = "ssh_key")]
    SshKey,
    #[serde(rename = "api_key")]
    #[ts(rename = "api_key")]
    ApiKey,
    Cookie,
    Certificate,
    Other,
}

impl VaultEntryType {
    pub fn as_str(&self) -> &'static str {
        match self {
            Self::Password => "password",
            Self::Token => "token",
            Self::SshKey => "ssh_key",
            Self::ApiKey => "api_key",
            Self::Cookie => "cookie",
            Self::Certificate => "certificate",
            Self::Other => "other",
        }
    }

    pub fn from_str(s: &str) -> Self {
        match s {
            "token" => Self::Token,
            "ssh_key" => Self::SshKey,
            "api_key" => Self::ApiKey,
            "cookie" => Self::Cookie,
            "certificate" => Self::Certificate,
            "other" => Self::Other,
            _ => Self::Password,
        }
    }
}
